### ISSUE TITLE
During training utterances are always processed sorted by duration

### ISSUE THREAD

**Issue by EgorLakomkin** [SeanNaren/deepspeech.pytorch#30]
As I remember in DeepSpeech paper the samples were sorted by duration only during the first epoch? Maybe it would make sense to switch from the sequential sampler to random sampler in AudioDataLoader after the first epoch?

**Comment by SeanNaren** [comment]
Agreed, I'm going to make sure that the random spike in memory doesn't occur in the pytorch version, this occurs in the torch lua version and caused some OOM issues...

**Comment by ryanleary** [comment]
I've had TED training most of the day at this point, and it looks to be suffering from overfitting/instability which may be caused by either this or the lack of batch norm (see plot).

<img width="1150" alt="screen shot 2017-04-26 at 10 39 39 pm" src="https://cloud.githubusercontent.com/assets/2212584/25466155/a0a462d4-2ad4-11e7-97a0-c904b86dd8a7.png">

@SeanNaren is the random spike you're talking about just caused by the random sample containing sequences that are too long in a given batch?

**Comment by SeanNaren** [comment]
Hmm thats really curious. I wonder how a large network does; could you try running the below command?

```python train.py --hidden_layers 5 --hidden_size 768```

Hopefully I'll be able to make changes to the PR allowing you to change RNNs; this might provide better results than using LSTMs. How long was you training time and did you run default parameters?

**Comment by EgorLakomkin** [comment]
In my experience, what really helped me with overfitting is the data augmentation. Random change of the tempo and loudness prevented overfitting. We could incorporate this into the DataLoader somehow, maybe through subclassing SpectrogramParser?

**Comment by SeanNaren** [comment]
@EgorLakomkin you're totally right, I need to get noise injection in ASAP. Will investigate further how this can be done!

**Comment by EgorLakomkin** [comment]
I used sox flags tempo <0.9 to 1.15> and gain <gain_value> and sometimes randomly append reverb

**Comment by ryanleary** [comment]
Kicked off a training with the larger parameters (was running default before).

Thoughts on setting up a wiki to track the performance progress on the various datasets as a function of hyperparameters and features as they're implemented?

**Comment by ryanleary** [comment]
With those params I'm getting an OOM runtime error on the gpu during eval interestingly.

**Comment by SeanNaren** [comment]
My guess is reducing the batch size will help to solve this however I'd like a script that could measure this for us...

**Comment by EgorLakomkin** [comment]
I got also OOM during the evaluation stage on LibriSpeech

**Comment by SeanNaren** [comment]
I've added a branch [here](https://github.com/SeanNaren/deepspeech.pytorch/tree/sorta-grad) that does random batches after the first epoch, however I do notice a 1gb VRAM jump when moving from the sequential sampler to the random sampler after the first epoch as was the case in the deepspeech.torch package.

I suspect it might have to do with the warp-ctc implementation but I'll have to verify this.

**Comment by ryanleary** [comment]
Any reason the sortagrad branch can't/shouldn't be merged in?

**Comment by SeanNaren** [comment]
There is a sharp increase in memory once sortagrad kicks in on the GPUs that I'm having a hard time figuring out...

**Comment by EgorLakomkin** [comment]
I also don't know the reason of a sharp increase, but maybe it has something to do with that there might be sequences with very different lengths in one batch?

**Comment by ryanleary** [comment]
@SeanNaren seems to me that the memory increase can be explained by the sequence length changes. It's likely that the memory usage will be similar to the last few minibatches of the first epoch since even if a single example is long, the full matrix becomes very large.

**Comment by ryanleary** [comment]
I think this can be closed with the merge of #83
